File: backend/app/pipeline/stats.py
"""Aggregate hit events into rallies and shot-type statistics."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Dict, List

from .hit_detection import HitEvent

ClassifyFn = Callable[[HitEvent, bool], str]


@dataclass
class Shot:
    t: float
    side: str
    shot_type: str
    rally_index: int


@dataclass
class AnalysisStats:
    shots: List[Shot]
    rally_count: int
    shot_counts: Dict[str, int]


def segment_rallies(hit_times: List[float], rally_gap_seconds: float) -> List[int]:
    """Assign each hit a rally index.

    A gap longer than `rally_gap_seconds` between consecutive hits starts a
    new rally (the ball went out of play / point ended).
    """
    if not hit_times:
        return []
    rally_indices = [0]
    for prev_t, t in zip(hit_times, hit_times[1:]):
        rally_indices.append(rally_indices[-1] + (1 if t - prev_t > rally_gap_seconds else 0))
    return rally_indices


def build_stats(hits: List[HitEvent], classify_fn: ClassifyFn, rally_gap_seconds: float) -> AnalysisStats:
    """Build shot-level stats from hit events.

    `classify_fn(hit, is_first_hit_of_rally)` is injected so this module
    stays independent of pose data - it only knows about hit timing.
    """
    rally_indices = segment_rallies([h.t for h in hits], rally_gap_seconds)
    shots: List[Shot] = []
    shot_counts: Dict[str, int] = {}

    for i, (hit, rally_idx) in enumerate(zip(hits, rally_indices)):
        is_first = i == 0 or rally_indices[i - 1] != rally_idx
        shot_type = classify_fn(hit, is_first)
        shots.append(Shot(t=hit.t, side=hit.side, shot_type=shot_type, rally_index=rally_idx))
        shot_counts[shot_type] = shot_counts.get(shot_type, 0) + 1

    rally_count = (rally_indices[-1] + 1) if rally_indices else 0
    return AnalysisStats(shots=shots, rally_count=rally_count, shot_counts=shot_counts)

File: docs/js/pipeline/stats.js
// Aggregate hit events into rallies and shot-type statistics.
// Mirrors backend/app/pipeline/stats.py.

/**
 * Assign each hit a rally index. A gap longer than `rallyGapSeconds`
 * between consecutive hits starts a new rally (the ball went out of
 * play / point ended).
 */
export function segmentRallies(hitTimes, rallyGapSeconds) {
  if (hitTimes.length === 0) return [];
  const rallyIndices = [0];
  for (let i = 1; i < hitTimes.length; i++) {
    const gap = hitTimes[i] - hitTimes[i - 1];
    rallyIndices.push(rallyIndices[rallyIndices.length - 1] + (gap > rallyGapSeconds ? 1 : 0));
  }
  return rallyIndices;
}

/**
 * Build shot-level stats from hit events.
 * `classifyFn(hit, isFirstHitOfRally)` is injected so this module stays
 * independent of pose data - it only knows about hit timing.
 */
export function buildStats(hits, classifyFn, rallyGapSeconds) {
  const rallyIndices = segmentRallies(
    hits.map((h) => h.t),
    rallyGapSeconds
  );
  const shots = [];
  const shotCounts = {};

  hits.forEach((hit, i) => {
    const rallyIndex = rallyIndices[i];
    const isFirst = i === 0 || rallyIndices[i - 1] !== rallyIndex;
    const shotType = classifyFn(hit, isFirst);
    shots.push({ t: hit.t, side: hit.side, shotType, rallyIndex });
    shotCounts[shotType] = (shotCounts[shotType] || 0) + 1;
  });

  const rallyCount = rallyIndices.length ? rallyIndices[rallyIndices.length - 1] + 1 : 0;
  return { shots, rallyCount, shotCounts };
}
